Makefile.rules

```diff
@@ -0,0 +1,162 @@
+	$(EXECUTABLE_LINK) -o $@ $(filter %.la, $^) $(filter %.lo, $^) $(LDFLAGS) $(LIBS)
+
+################################################################################
+# Installation rules
+
+install: install-headers install-pkgconfig install-static install-shared
+
+install-headers:
+	for file in $(HEADERS) ; do \
+		$(INSTALL) -m 644 -v -D $$file $(DESTDIR)$(includedir)/$$file ; \
+	done
+
+install-pkgconfig:
+	mkdir -p $(DESTDIR)$(libdir)/pkgconfig/
+	for file in $(PKGCONFIG_FILES) ; do \
+		$(INSTALL) -m 644 -v -D $$file $(DESTDIR)$(libdir)/pkgconfig/ ; \
+	done
+
+install-static: $(STATIC_LIBRARIES)
+	mkdir -p $(DESTDIR)$(libdir)/
+	for file in $(filter %.a,$(STATIC_LIBRARIES)) ; do\
+		$(INSTALL) -m 644 -v $$file $(DESTDIR)$(libdir)/ ; \
+	done
+
+install-shared: $(patsubst %.so,install-shared-%,$(SHARED_LIBRARIES))
+install-shared-%: BASENAME=$(patsubst install-shared-%,%,$@)
+install-shared-%:
+	mkdir -p $(DESTDIR)$(libdir)/
+	$(INSTALL) -m 755 -v $(BASENAME).so.$($(BASENAME)_VERSION) $(DESTDIR)$(libdir)/
+	$(RM) -rf $(DESTDIR)$(libdir)/$(BASENAME).so $(DESTDIR)$(libdir)/$(BASENAME).so.$($(BASENAME)_SOVERSION)
+	ln -sf $(BASENAME).so.$($(BASENAME)_VERSION) $(DESTDIR)$(libdir)/$(BASENAME).so
+	ln -sf $(BASENAME).so.$($(BASENAME)_VERSION) $(DESTDIR)$(libdir)/$(BASENAME).so.$($(BASENAME)_SOVERSION)
+
+################################################################################
+# Pseudo-rules for maintaining development build tree
+
+clean:
+	find $(BUILD_OUT_PREFIX). -name "*.[ao]" -o -name "*.l[ao]" -o -name "*.so" -o -name "*.so.*" -o -name "*.gcda" -o -name "*.gcno" | xargs rm -f
+	rm -f $(GENERATED_FILES)
+
+distclean: clean depclean
+	rm -rf config.log config.status include/tscb/config Makefile .depend autom4te.cache $(PKGCONFIG_FILES)
+
+maintainerclean: distclean
+	rm -f configure
+
+# dependency rules
+
+$(BUILD_OUT_PREFIX).dep/%.la.d: %.cpp
+	@mkdir -p $(dir $@)
+	@$(CXX) -MM $(CXXFLAGS) $(CPPFLAGS) -MT $(patsubst %.cpp, $(BUILD_OUT_PREFIX)%.la, $<) -MP -MF $@ $<
+	@echo MAKEDEP $<
+
+$(BUILD_OUT_PREFIX).dep/%.lo.d: %.cpp
+	@mkdir -p $(dir $@)
+	@$(CXX) -MM -DPIC -fPIC $(CXXFLAGS) $(CPPFLAGS) -MT $(patsubst %.cpp, $(BUILD_OUT_PREFIX)%.lo, $<) -MP -MF $@ $<
+
+depclean:
+	rm -rf $(BUILD_OUT_PREFIX).dep
+
+DEPEND = $(patsubst %.cpp, $(BUILD_OUT_PREFIX).dep/%.la.d, $(SOURCES)) $(patsubst %.cpp, $(BUILD_OUT_PREFIX).dep/%.lo.d, $(SOURCES))
+depend: $(DEPEND)
+ifeq ($(shell if [ -e $(BUILD_OUT_PREFIX).dep ] ; then echo yes ; fi),yes)
+-include $(DEPEND)
+endif
+
+echo-depends:
+	echo $(DEPEND)
+
+################################################################################
+# Unit testing rules
+
+SOURCES+=$(patsubst %, %.cpp, $(TESTS))
+RUNTESTS=$(patsubst %, run-%, $(TESTS))
+VALGRINDTESTS=$(patsubst %, valgrind-%, $(TESTS))
+
+$(TESTS): % : %.la
+$(TESTS): LDFLAGS += -pthread
+
+$(RUNTESTS): run-% : %
+	$^
+
+$(VALGRINDTESTS): valgrind-% : %
+	valgrind --leak-check=full --error-exitcode=1 $^
+
+check: $(RUNTESTS)
```

Comment:
This compiles each test individually and immediately runs it afterwards. The test-runner executable does not seem to be used at all, even you modified it in this PR.

Reply:
The test-runner is still used as the "main" program for each test. But now it is for each test file individually (and each test file may still register multiple tests inside it).